pkg/util/crypt/crypt_dev.go

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2018-2019, Sylabs Inc. All rights reserved.
+// This software is licensed under a 3-clause BSD license. Please consult the
+// LICENSE.md file distributed with the sources of this project regarding your
+// rights to use or distribute this software.
+
+package crypt
+
+import (
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+
+	"github.com/sylabs/singularity/internal/pkg/sylog"
+	"github.com/sylabs/singularity/pkg/util/fs/lock"
+	"github.com/sylabs/singularity/pkg/util/loop"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+// Device is
+type Device struct{}
+
+// createLoop attaches the file to the next available loop device and
+// sets the sizelimit on it
+func createLoop(file *os.File, offset, size uint64) (string, error) {
+	loopDev := &loop.Device{
+		MaxLoopDevices: 256,
+		Shared:         true,
+		Info: &loop.Info64{
+			SizeLimit: size,
+			Offset:    offset,
+			Flags:     loop.FlagsAutoClear,
+		},
+	}
+	idx := 0
+	if err := loopDev.AttachFromFile(file, os.O_RDWR, &idx); err != nil {
+		return "", fmt.Errorf("failed to attach image %s: %s", file.Name(), err)
+	}
+	return fmt.Sprintf("/dev/loop%d", idx), nil
+}
+
+// CloseCryptDevice closes the crypt device
+func (crypt *Device) CloseCryptDevice(path string) error {
+
+	cmd := exec.Command("/sbin/cryptsetup", "luksClose", path)
+	cmd.SysProcAttr = &syscall.SysProcAttr{}
+	cmd.SysProcAttr.Credential = &syscall.Credential{Uid: 0, Gid: 0}
+	fd, err := lock.Exclusive("/dev/mapper")
+	if err != nil {
+		return err
+	}
+	err = cmd.Run()
+	if err != nil {
+		sylog.Debugf("Unable to delete the crypt device %s", err)
+		return err
+	}
+	err = lock.Release(fd)
+	if err != nil {
+		sylog.Debugf("Unable to release the lock on /dev/mapper")
+		return err
+	}
+
+	return nil
+}
+
+// FormatCryptDevice allocates a loop device, encrypts, and returns the loop device name, and encrypted device name
+func (crypt *Device) FormatCryptDevice(path string) (string, string, error) {
+
+	// Read the password from terminal
+	fmt.Print("Enter the password to encrypt the filesystem: ")
+	password, err := terminal.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		sylog.Fatalf("Error parsing the password: %s", err)
+	}
+	input := string(password)
+
+	fmt.Print("\nConfirm the password: ")
+	password2, err := terminal.ReadPassword(int(syscall.Stdin))
+	if err != nil {
+		sylog.Fatalf("Error parsing the password: %s", err)
+	}
+	input2 := string(password2)
+	fmt.Println()
+
+	if input != input2 {
+		return "", "", errors.New("Passwords don't match")
+	}
+
+	squashF, err := os.Stat(path)
+	if err != nil {
+		sylog.Debugf("Error acquiring size of %s", path)
+		return "", "", err
```

Comment:
I think this should also return some debugging like the above line

```suggestion
		return "", "", fmt.Errorf("failed getting size of: %s: ", path, err)
```